pyro/ops/gaussian.py

```diff
@@ -251,33 +277,49 @@ class AffineNormal:
     """
     def __init__(self, matrix, loc, scale):
         assert loc.shape == scale.shape
+        assert matrix.shape[:-2] == loc.shape[:-1]
+        assert matrix.size(-1) == loc.size(-1)
         self.matrix = matrix
         self.loc = loc
         self.scale = scale
+        self._gaussian = None
 
-    @property
+    @lazy_property
     def batch_shape(self):
         return self.matrix.shape[:-2]
 
     def condition(self, value):
+        if value.size(-1) == self.loc.size(-1):
+            prec_sqrt = self.matrix / self.scale.unsqueeze(-2)
+            precision = matmul(prec_sqrt, prec_sqrt.transpose(-1, -2))
+            delta = (value - self.loc) / self.scale
+            info_vec = matvecmul(prec_sqrt, delta)
+            log_normalizer = (-0.5 * self.loc.size(-1) * math.log(2 * math.pi)
+                              - 0.5 * delta.pow(2).sum(-1) - self.scale.log().sum(-1))
+            return Gaussian(log_normalizer, info_vec, precision)
+        else:
+            return self.to_gaussian().condition(value)
+
+    def left_condition(self, value):
         """
-        Condition on a ``Y`` value.
-
-        :param torch.Tensor value: A value of ``Y``.
-        :return Gaussian: A gaussian likelihood over ``X``.
+        If ``value.size(-1) == x_dim``, this returns a Normal distribution with
+        ``event_dim=1``. After applying this method, the cost to draw a sample is
+        ``O(y_dim)`` instead of ``O(y_dim ** 3)``.
         """
-        assert value.size(-1) == self.loc.size(-1)
-        prec_sqrt = self.matrix / self.scale.unsqueeze(-2)
-        precision = matmul(prec_sqrt, prec_sqrt.transpose(-1, -2))
-        delta = (value - self.loc) / self.scale
-        info_vec = matvecmul(prec_sqrt, delta)
-        log_normalizer = (-0.5 * self.loc.size(-1) * math.log(2 * math.pi)
-                          - 0.5 * delta.pow(2).sum(-1) - self.scale.log().sum(-1))
-        return Gaussian(log_normalizer, info_vec, precision)
+        if value.size(-1) == self.matrix.size(-2):
+            loc = matvecmul(self.matrix.transpose(-1, -2), value) + self.loc
+            return torch.distributions.Independent(
+                torch.distributions.Normal(loc, scale=self.scale), 1)
```

Comment:
This makes the output a bit inconsistent. Should we merge `rsample` with `left_condition`, e.g. `rsample(sample_shape=..., left_condition=None)`? I am open to any suggestions for having a consistent output.

Reply:
I see, the output type depends on input dims...

I think it's fine to implement a fused method `.left_condition_rsample(value, sample_shape=torch.Size())`.

Does the following alternative make sense: we could return a `Gaussian` here with event_dim=1 and an extra batch dimension, then update `Gaussian.rsample()` to handle the special case of `evet_dim=1`, as you did with Funsor. I guess in this case the type would be consistent but the `batch_shape` would be inconsistent?

Reply:
I think we can implement a fused method. Supporting `event_dim=1` in Gaussian would require a lot of ops. On second thought, I think we can return an AffineNormal here and add a `rsample` method to `AffineNormal`. Like Gaussian, which only supports .rsample for full rank precision, here we support `.rsample` for empty matrix. WDYT?

Reply:
sounds reasonable, would `.matrix is None` correspond to an identity matrix?

Reply:
It is just simply a zeros/empty matrix with 0 size: `.matrix = torch.zeros(0, loc.size(-1))`.